File: race-worker/src/safety.js
// Photosensitivity guard for a building-sized display.
//
// WCAG 2.3.1: nothing may flash more than 3 times in any 1 second. A flash is a pair of
// opposing changes in relative luminance of 10% or more where the darker state is below 0.8.
// A whole facade is a very large visual field, so the guard enforces a stricter budget, on
// every window and on the facade average:
//   1. Slew: each R/G/B moves at most MAX_STEP per frame (hard cuts become short fades).
//   2. Reversal budget: a change that would exceed GUARD_FLASHES_PER_SECOND is held back.
// Scenes should still avoid flashy effects instead of leaning on the guard.

export const MAX_FLASHES_PER_SECOND = 3;
export const GUARD_FLASHES_PER_SECOND = 2;
const FLASH_DELTA = 0.1;
const DARK_BELOW = 0.8;
const RAMP_SECONDS = 0.3; // a 0 -> 255 fade takes at least this long at the reference fps
const REFERENCE_FPS = 30;

const LINEAR = Array.from({ length: 256 }, (_, v) => {
  const s = v / 255;
  return s <= 0.04045 ? s / 12.92 : ((s + 0.055) / 1.055) ** 2.4;
});

export function luminance([r, g, b]) {
  return 0.2126 * LINEAR[r] + 0.7152 * LINEAR[g] + 0.0722 * LINEAR[b];
}

/** Counts WCAG transitions (half-flashes) on one luminance signal, with timestamps in ms. */
export class FlashTracker {
  constructor(value = 0) {
    this.lo = this.hi = value;
    this.direction = 0;
    this.marks = [];
  }

  flip(v) {
    const hi = Math.max(this.hi, v);
    const lo = Math.min(this.lo, v);
    if (this.direction >= 0 && hi - v >= FLASH_DELTA && v < DARK_BELOW) return -1;
    if (this.direction <= 0 && v - lo >= FLASH_DELTA && lo < DARK_BELOW) return 1;
    return 0;
  }

  recent(now) {
    while (this.marks.length && this.marks[0] <= now - 1000) this.marks.shift();
    return this.marks.length;
  }

  allows(v, now, perSecond) {
    return this.flip(v) === 0 || this.recent(now) < 2 * perSecond;
  }

  commit(v, now) {
    const flip = this.flip(v);
    this.hi = Math.max(this.hi, v);
    this.lo = Math.min(this.lo, v);
    if (flip !== 0) {
      this.marks.push(now);
      this.direction = flip;
      this.hi = this.lo = v;
    }
    return this.recent(now);
  }
}

const facadeLuminance = (grid) => {
  let sum = 0;
  let n = 0;
  for (const row of grid) for (const px of row) (sum += luminance(px)), n++;
  return sum / n;
};

export class FlashGuard {
  constructor({ perSecond = GUARD_FLASHES_PER_SECOND } = {}) {
    this.perSecond = perSecond;
    this.shown = null;
    this.lastAt = null;
    this.trackers = null;
    this.facade = null;
  }

  /** Returns the grid that is safe to show at time `now` (ms). Frame spacing may vary. */
  filter(grid, now) {
    if (!this.shown) {
      this.shown = grid.map((row) => row.map((px) => [...px]));
      this.trackers = grid.map((row) => row.map((px) => new FlashTracker(luminance(px))));
      this.facade = new FlashTracker(facadeLuminance(grid));
      this.lastAt = now;
      return this.shown;
    }
    const dt = Math.max(1, now - this.lastAt);
    this.lastAt = now;
    const step = Math.max(1, Math.round((255 / (RAMP_SECONDS * REFERENCE_FPS)) * (dt / (1000 / REFERENCE_FPS))));

    const out = grid.map((row, r) =>
      row.map((want, c) => {
        const prev = this.shown[r][c];
        let next = prev.map((p, i) => p + Math.max(-step, Math.min(step, want[i] - p)));
        const tracker = this.trackers[r][c];
        if (!tracker.allows(luminance(next), now, this.perSecond)) next = prev;
        tracker.commit(luminance(next), now);
        return next;
      }),
    );

    let shown = out;
    if (!this.facade.allows(facadeLuminance(out), now, this.perSecond)) {
      shown = this.shown; // the whole facade would flash: hold the last frame
      shown.forEach((row, r) => row.forEach((px, c) => this.trackers[r][c].commit(luminance(px), now)));
    }
    this.facade.commit(facadeLuminance(shown), now);
    this.shown = shown;
    return shown;
  }
}

/** Worst flashes-per-second in any 1 s window: { rate, where } (where = "r,c" or "facade"). */
export function worstFlashRate(frames, fps) {
  const trackers = new Map();
  let worst = { rate: 0, where: null };
  frames.forEach((grid, i) => {
    const now = (i * 1000) / fps;
    const values = [["facade", facadeLuminance(grid)]];
    grid.forEach((row, r) => row.forEach((px, c) => values.push([`${r},${c}`, luminance(px)])));
    for (const [where, v] of values) {
      if (!trackers.has(where)) trackers.set(where, new FlashTracker(v));
      const rate = trackers.get(where).commit(v, now) / 2;
      if (rate > worst.rate) worst = { rate, where };
    }
  });
  return worst;
}

File: tamagochi-mascots-python/tower/safety.py
"""Photosensitivity guard for a building-sized display.

WCAG 2.3.1: content must not flash more than 3 times in any 1 second. A flash is a pair of
opposing changes in relative luminance of 10% or more, where the darker state is below 0.8.
A whole facade is a very large visual field, so we enforce a stricter budget and enforce it
in code, on every window and on the facade as a whole:

1. Slew: a window's R, G, B may move at most MAX_STEP per frame, so hard cuts become soft fades.
2. Reversal budget: each window (and the facade average) is tracked with the WCAG flash
   definition; a frame change that would push it past GUARD_FLASHES_PER_SECOND is held back
   until the budget frees up.

Apps should still avoid flashy effects instead of leaning on the guard; tests/test_safety.py
runs every shipped animation through worst_flash_rate().
"""

from collections import deque

from tower.display import COLS, MAX_FPS, ROWS, Frame

MAX_FLASHES_PER_SECOND = 3     # WCAG 2.3.1 general flash threshold
GUARD_FLASHES_PER_SECOND = 2   # what the guard allows - margin under the threshold
RAMP_SECONDS = 0.3             # a 0 -> 255 fade takes at least this long
MAX_STEP = max(1, round(255 / (RAMP_SECONDS * MAX_FPS)))
FLASH_DELTA = 0.1
DARK_BELOW = 0.8


def _linear(v):
    v /= 255
    return v / 12.92 if v <= 0.04045 else ((v + 0.055) / 1.055) ** 2.4


_LIN = [_linear(v) for v in range(256)]


def luminance(rgb):
    """WCAG relative luminance, 0..1."""
    return 0.2126 * _LIN[rgb[0]] + 0.7152 * _LIN[rgb[1]] + 0.0722 * _LIN[rgb[2]]


class FlashTracker:
    """Counts WCAG transitions (half-flashes) on one luminance signal."""

    def __init__(self, v=0.0, fps=MAX_FPS):
        self.fps = fps
        self.lo = self.hi = v
        self.direction = 0
        self.marks = deque()

    def _flip(self, v):
        """New direction if luminance v would count as a transition, else None."""
        hi, lo = max(self.hi, v), min(self.lo, v)
        if self.direction >= 0 and hi - v >= FLASH_DELTA and v < DARK_BELOW:
            return -1
        if self.direction <= 0 and v - lo >= FLASH_DELTA and lo < DARK_BELOW:
            return 1
        return None

    def recent(self, i):
        while self.marks and self.marks[0] <= i - self.fps:
            self.marks.popleft()
        return len(self.marks)

    def allows(self, v, i, per_second):
        return self._flip(v) is None or self.recent(i) < 2 * per_second

    def commit(self, v, i):
        flip = self._flip(v)
        self.hi, self.lo = max(self.hi, v), min(self.lo, v)
        if flip is not None:
            self.marks.append(i)
            self.direction = flip
            self.hi = self.lo = v
        return self.recent(i)


class FlashGuard:
    def __init__(self, display=None, max_step=MAX_STEP, per_second=GUARD_FLASHES_PER_SECOND):
        self.display = display
        self.max_step = max_step
        self.per_second = per_second
        self.shown = Frame()
        self.trackers = [[FlashTracker() for _ in range(COLS)] for _ in range(ROWS)]
        self.facade = FlashTracker()
        self.i = 0

    def filter(self, frame):
        step, i = self.max_step, self.i
        out = Frame()
        for r in range(ROWS):
            for c in range(COLS):
                prev, want = self.shown.px[r][c], frame.px[r][c]
                nxt = tuple(p + max(-step, min(step, w - p)) for p, w in zip(prev, want))
                tracker = self.trackers[r][c]
                if not tracker.allows(luminance(nxt), i, self.per_second):
                    nxt = prev
                tracker.commit(luminance(nxt), i)
                out.px[r][c] = nxt
        mean = sum(luminance(p) for row in out.px for p in row) / (ROWS * COLS)
        if not self.facade.allows(mean, i, self.per_second):
            out = self.shown  # the whole facade would flash: hold the last frame
            mean = sum(luminance(p) for row in out.px for p in row) / (ROWS * COLS)
            for r in range(ROWS):
                for c in range(COLS):
                    self.trackers[r][c].commit(luminance(out.px[r][c]), i)
        self.facade.commit(mean, i)
        self.shown = out
        self.i += 1
        return out

    def send(self, frame):
        self.display.send(self.filter(frame))


def worst_flash_rate(frames, fps=MAX_FPS):
    """Highest flashes per second in any 1 s window, as (rate, where); where is (row, col) or "facade"."""
    trackers = {(r, c): None for r in range(ROWS) for c in range(COLS)}
    trackers["facade"] = None
    worst = (0.0, None)
    for i, f in enumerate(frames):
        values = {(r, c): luminance(f.px[r][c]) for r in range(ROWS) for c in range(COLS)}
        values["facade"] = sum(values.values()) / (ROWS * COLS)
        for where, v in values.items():
            if trackers[where] is None:
                trackers[where] = FlashTracker(v, fps)
            rate = trackers[where].commit(v, i) / 2
            if rate > worst[0]:
                worst = (rate, where)
    return worst
